### DISCUSSION TITLE
there is any way to change pydantic body base incoming request?

### DISCUSSION THREAD

**Discussion by zedrood** [fastapi/fastapi#6880] [Questions] [answered]
in django rest framework we use `get_serializer_class` in `GenericAPIView` to change serializer base incoming request .
I use multiple type user and base on each user needs different pydantic body. for solve and handling  i use multiple route. there is any way to change pydantic body base incoming request?

**Comment by ycd** [comment]
You can use something like this


```python
class A(BaseModel):
    b: Union[SomeModelA, SomeModelB]

    @validator("b", pre=True)
    def validate_b(cls, value):
        if isinstance(value, dict):
            someField = value.get("someInnerField")
            if someField == "abc":
                return SomeModelA(**value)
            elif someField == "def":
                return SomeModelB(**value)
```

**Comment by zedrood** [comment]
thanks for answer @ycd  but problem is not solve each user has own permission on request and i use custom pydantic body for each user. any user can send custom pydantic body and we no that user can access on  pydantic body. can check permission user before check body and set own custom pydantic body ?

**Comment by zedrood** [comment]
let me explain something like this in drf 
```python
class ChangeSerializerBaseTypeUserView(GenericViewSet):
        ...
	
	def get_serializer_class(self):
		if self.request.user.type == 1:
		    self.serializer_class = OneSerializer
		elif self.request.user.type == 2:
		    self.serializer_class = TwoSerializer
		else:
		    self.serializer_class = OtherSerializer
		return self.serializer_class
	
	...
```
there is somthing like that in fastapi ? 
i use Depends for check permission and get current user.

**Comment by insomnes** [comment]
@zedrood hello!
You can try to use `def handler(user: Union[UserModelA, UserModelB, UserModelC]):` if you have different user models which have some distinct fields.

**Comment by ycd** [comment]
@insomnes as far as i understand from the example he gave above, the fields are the same, but he wants to use different serializer for same fields with different types. 

>  i use custom pydantic body for each user. any user can send custom pydantic body and we no that user can access on pydantic body. can check permission user before check body and set own custom pydantic body ?

Can you rephrase this?

**Comment by zedrood** [comment]
Thanks for your answer two people @insomnes , @ycd
let me explain with example for clear what i want
i two type user and two different serializer 
```python
class Test1Serializer(BaseModel):
    x: str
    y:int 
    z: float 


class Test2Serializer(BaseModel):
    t: str
    n:str 
    u: str
```
for use two user and serializer i must be two different route like this
```python
@router.post("/test1/")
async def test1(
    test_serializer: Test1Serializer = Body(...),
    current_user = Depends(get_user1),
):
    ...


@router.post("/test2/")
async def test2(
    test_serializer: Test2Serializer = Body(...),
    current_user = Depends(get_user2),
):
    ...
```
but i want merge two upper route in one route. before client send body i want to check permission user and set custom serializer (pydantic) body. for now i use this below code but the client can send any of the serializers and these serializers are own validation and client base on own permission not allow to send other serializers.
```python
@router.post("/test/")
async def test(
    test_serializer: Union[Test1Serializer, Test2Serializer] = Body(...),
    current_user = Depends(get_user1_or_user2),
):
    if current_user.type == 1 and type(test_serializer) == Test1Serializer :
        ...
    elif current_user.type == 2 and type(test_serializer) == Test2Serializer :
        ...
```

**Comment by zedrood** [comment, marked as answer]
i resolve this problem with this below code 
```python
from fastapi.exception_handlers import RequestValidationError
from pydantic import ValidationError

@router.post("/test/")
async def test(
    test_serializer = Body(...),
    current_user = Depends(get_user1_or_user2),
):
	
    if current_user.type == 1:
    	try:
    		data = Test1Serializer(**test_serializer).dict()
    	except ValidationError as ex:
    		raise RequestValidationError(errors=ex.raw_errors, body=test_serializer)
        
    
    elif current_user.type == 2:
    	try:
    		data = Test2Serializer(**test_serializer).dict()
    	except ValidationError as ex:
    		raise RequestValidationError(errors=ex.raw_errors, body=test_serializer)
   
     ...
```

**Comment by tiangolo** [comment]
Thanks for the help here everyone! 👏  🙇 

Thanks for reporting back and closing the issue @zedrood  👍
